packages/flutter/lib/src/material/dropdown.dart

```diff
@@ -1036,10 +1031,27 @@ class _DropdownButtonState<T> extends State<DropdownButton<T>> with WidgetsBindi
 
   bool get _enabled => widget.items != null && widget.items.isNotEmpty && widget.onChanged != null;
 
+  Orientation _getOrientation(BuildContext context) {
+    Orientation result = MediaQuery.of(context, nullOk: true)?.orientation;
+    if (result == null) {
+      // If there's no MediaQuery, then use the window aspect to determine
+      // orientation.
+      final Size size = window.physicalSize;
+      result = size.width > size.height ? Orientation.landscape : Orientation.portrait;
+    }
+    return result;
+  }
+
   @override
   Widget build(BuildContext context) {
     assert(debugCheckHasMaterial(context));
     assert(debugCheckHasMaterialLocalizations(context));
+    final Orientation newOrientation = _getOrientation(context);
+    _lastOrientation ??= newOrientation;
+    if (newOrientation != _lastOrientation) {
+      _removeDropdownRoute();
```

Comment:
This shouldn't be called from build. I am surprised that this is actually working and doesn't throw an "setState during build is illegal" exception.

Can you not just check in the didChangeMetrics if the orientation has changed from the last time and if it did remove the route there?

Reply:
Actually, looks like we have special code in place to allow removing an overlay during build: https://github.com/flutter/flutter/blame/166d422b05e69ea140fcdc8968282b883aad0af5/packages/flutter/lib/src/widgets/overlay.dart#L128

@Hixie Do you know why we allow this?